File: lfs-client-sdk/python/lfs_sdk/envelope.py
from __future__ import annotations

from dataclasses import dataclass
import json
from typing import Any, Dict


@dataclass
class LfsEnvelope:
    kfs_lfs: int
    bucket: str
    key: str
    size: int
    sha256: str
    checksum: str | None = None
    checksum_alg: str | None = None
    content_type: str | None = None
    original_headers: Dict[str, str] | None = None
    created_at: str | None = None
    proxy_id: str | None = None


def is_lfs_envelope(value: bytes | None) -> bool:
    if not value or len(value) < 15:
        return False
    if value[:1] != b"{":
        return False
    prefix = value[:50].decode("utf-8", errors="ignore")
    return "\"kfs_lfs\"" in prefix


def decode_envelope(value: bytes) -> LfsEnvelope:
    payload = json.loads(value.decode("utf-8"))
    if not payload.get("kfs_lfs") or not payload.get("bucket") or not payload.get("key") or not payload.get("sha256"):
        raise ValueError("invalid envelope: missing required fields")
    return LfsEnvelope(**payload)

File: lfs-client-sdk/js/src/envelope.ts
export interface LfsEnvelope {
  kfs_lfs: number;
  bucket: string;
  key: string;
  size: number;
  sha256: string;
  checksum?: string;
  checksum_alg?: string;
  content_type?: string;
  original_headers?: Record<string, string>;
  created_at?: string;
  proxy_id?: string;
}

export function isLfsEnvelope(value: Uint8Array | null | undefined): boolean {
  if (!value || value.length < 15) return false;
  if (value[0] !== 123) return false;
  const prefix = new TextDecoder().decode(value.slice(0, Math.min(50, value.length)));
  return prefix.includes('"kfs_lfs"');
}

export function decodeEnvelope(value: Uint8Array): LfsEnvelope {
  const env = JSON.parse(new TextDecoder().decode(value)) as LfsEnvelope;
  if (!env.kfs_lfs || !env.bucket || !env.key || !env.sha256) {
    throw new Error('invalid envelope: missing required fields');
  }
  return env;
}
